### ISSUE TITLE
bridge connects to remote EMQX even though server name does not match certificate

### ISSUE THREAD

**Issue by otengler** [emqx/emqx#5277]
My EMQX version is 4.2.0 which I run on Windows.
I use the EMQX bridge plugin to connect to another EMQX broker. This works flawlessly. But my problem is that it even works when I think that it shouldn't because the "cacertfile" in my bridge configuration doesn't belong to the server I am connecting to.
In my bridge config below I am connecting to "remotebroker" but the cacertfile is issued for "localbroker". 
I expected this to fail. Is there some configuration setting that I am missing? Or is my expectation wrong?
The same happens when I upgrade to 4.3.5.

This is my bridge config:
> bridge.mqtt.bridge1.address = remotebroker:8883
> bridge.mqtt.bridge1.proto_ver = mqttv4
> bridge.mqtt.bridge1.start_type = auto
> bridge.mqtt.bridge1.bridge_mode = true
> bridge.mqtt.bridge1.clientid = bridge1-8e7ace21-c551-42a4-921d-3ee73517c9e6
> bridge.mqtt.bridge1.clean_start = true
> bridge.mqtt.bridge1.username = <user>
> bridge.mqtt.bridge1.password = <password>
> bridge.mqtt.bridge1.forwards = # omitted
> bridge.mqtt.bridge1.subscription.1.topic = # omitted
> bridge.mqtt.bridge1.subscription.1.qos = 1
> bridge.mqtt.bridge1.ssl = on
> bridge.mqtt.bridge1.cacertfile = etc/certs/localbroker.pem
> bridge.mqtt.bridge1.ciphers = ECDHE-ECDSA-AES256-GCM-SHA384,ECDHE-RSA-AES256-GCM-SHA384,ECDHE-ECDSA-AES256-SHA384,ECDHE-RSA-AES256-SHA384,ECDHE-ECDSA-DES-CBC3-SHA,ECDH-ECDSA-AES256-GCM-SHA384,ECDH-RSA-AES256-GCM-SHA384,ECDH-ECDSA-AES256-SHA384,ECDH-RSA-AES256-SHA384,DHE-DSS-AES256-GCM-SHA384,DHE-DSS-AES256-SHA256,AES256-GCM-SHA384,AES256-SHA256,ECDHE-ECDSA-AES128-GCM-SHA256,ECDHE-RSA-AES128-GCM-SHA256,ECDHE-ECDSA-AES128-SHA256,ECDHE-RSA-AES128-SHA256,ECDH-ECDSA-AES128-GCM-SHA256,ECDH-RSA-AES128-GCM-SHA256,ECDH-ECDSA-AES128-SHA256,ECDH-RSA-AES128-SHA256,DHE-DSS-AES128-GCM-SHA256,DHE-DSS-AES128-SHA256,AES128-GCM-SHA256,AES128-SHA256,ECDHE-ECDSA-AES256-SHA,ECDHE-RSA-AES256-SHA,DHE-DSS-AES256-SHA,ECDH-ECDSA-AES256-SHA,ECDH-RSA-AES256-SHA,AES256-SHA,ECDHE-ECDSA-AES128-SHA,ECDHE-RSA-AES128-SHA,DHE-DSS-AES128-SHA,ECDH-ECDSA-AES128-SHA,ECDH-RSA-AES128-SHA,AES128-SHA
> bridge.mqtt.bridge1.keepalive = 60s
> bridge.mqtt.bridge1.tls_versions = tlsv1.2,tlsv1.1,tlsv1
> bridge.mqtt.bridge1.reconnect_interval = 30s
> bridge.mqtt.bridge1.retry_interval = 5s
> bridge.mqtt.bridge1.batch_size = 32
> bridge.mqtt.bridge1.max_inflight_size = 32
> bridge.mqtt.bridge1.queue.replayq_dir = data/bridges/bridge1/
> bridge.mqtt.bridge1.queue.replayq_seg_bytes = 1MB
> bridge.mqtt.bridge1.queue.max_total_size = 10MB

Thanks for your time.

**Comment by qzhuyan** [comment]
thanks for reporting,

my gut feeling is missing verify_peer in the SSL option in the bridge plugin. potential to be a new feature for emqtt https://github.com/emqx/emqtt

**Comment by otengler** [comment]
@qzhuyan You mean similar to what has been added to the auth-http plugin?

Excerpt from emqx_auth_http.conf:
> \#\# In mode verify_none the default behavior is to allow all x509-path
> \#\# validation errors.
> \#\#
> \#\# Value: true | false
> \#\# **auth.http.ssl.verify = false**
>
> \#\# If not specified, the server's names returned in server's certificate is validated against
> \#\# what's provided `auth.http.auth_req.url` config's host part.
> \#\# Setting to 'disable' will make EMQ X ignore unmatched server names.
> \#\# If set with a host name, the server's names returned in server's certificate is validated
> \#\# against this value.
> \#\#
> \#\# Value: String | disable
> \#\# **auth.http.ssl.server_name_indication = disable**

Should I open a feature request for this? 

Regards,
Oliver

**Comment by qzhuyan** [comment]
@otengler 
In your case, something like the following should work:
```
bridge.mqtt.bridge1.verify=true
bridge.mqtt.bridge1.server_name_indication=some.hostname
```

note, I just checked that in version 4.3.5, Let me know if it doesn't work for 4.2.0

**Comment by otengler** [comment]
@qzhuyan 
Thank you very much for the hint. I tried to replicate your results on 4.3.5 but as soon as I specify "verify" or "server_name_indication" the bridge plugin refuses to load:

> [error] You've tried to set bridge.mqtt.bridge1.server_name_indication, but there is no setting with that name.
> [error] Did you mean one of these?
> [error] bridge.mqtt.$name.username
> [error] bridge.mqtt.$name.clientid
> [error] bridge.mqtt.$name.forward_mountpoint
> [error] You've tried to set bridge.mqtt.bridge1.verify, but there is no setting with that name.
> [error] Did you mean one of these?
> [error] bridge.mqtt.$name.certfile
> [error] bridge.mqtt.$name.keyfile
> [error] bridge.mqtt.$name.address

I checked that I'm really running 4.3.5 by subscribing to $SYS:
> $SYS/brokers/emqx@127.0.0.1/version  QoS: 0
> 4.3.5

**Comment by qzhuyan** [comment]
@otengler oh sorry I miss that the plugin config schema file needs an update. I will fix it.

**Comment by qzhuyan** [comment]
@otengler I have a draft patch for you if you want to have a test.
https://github.com/qzhuyan/emqx/commit/3c7acdefb4f8e3f402b3b4543cc97c4801c617ad

**Comment by otengler** [comment]
@qzhuyan Wow, that has been a really fast fix. Thanks a lot. I try to test this during the next days. This change will be part of a future release, right?

**Comment by otengler** [comment]
@qzhuyan Sorry, I'm not able to verify your changes. I tried to build your branch on my Windows machine but it fails after a while with:
```
Compiling emqx_bridge_mqtt
[relup_helper] running gen_appups
operable program or batch file.not recognized as an internal or external command,
...
←[1mUncaught error in rebar_core. Run with DIAGNOSTIC=1 to see stacktrace or consult rebar3.crashdump
```

I also tried with DIAGNOSTIC=1 on but it didn't get me any further. Probably something is wrong with my environment. I think it is OK to close this ticket.

**Comment by qzhuyan** [comment]
@otengler sorry for the late reply, I just come back from vacation.

Yes. the fix would be included in the next 4.3 release.
